### PR TITLE
SCC: Add vectorisation annotations in SCCRevector and translate in SCCAnnotate 

### PR DESCRIPTION
**Pull request by mlange05**
~_Note: Just testing for now..._~

This PR changes the way we add OpenACC annotations in the SCC pipelines and generally refactors `SCCAnnotate` and `SCCRevector`. The key change is that we now insert Loki-specific annotations for vector-loops, driver loops, routine annotations and vector reductions in `SCCRevector` and then let `SCCAnnotate` translate this into OpenACC directives. This is in preparation for the addition of an additional re-vectorisation scheme and the eventual support for OpenMP-offload in the SCC pipelines, as it will allow `SCCAnnotate` to easily switch between directive flavours.

There's also quite a bit of refactoring and clean-up, in particular in the upper control methods of `SCCAnnotate`, as well as the utility method `check_routine_pragmas`. Overall I hope this is now a bit cleaner and better structured for future development.

In more detail:
* `wrap_vector_section` has been made a standalone routine (will be re-used in alternative re-vector scheme).
* `SCCRevector` now annotates vector and sequential loops with respective `!$loki loop` pragmas, as well as marking kernel routines `!%loki routine seq|vector` and adding `!$loki loop vector-reductions` for vector loops in marked vector-reduction regions.
* `SCCRevector` also re-uses some of these utilities when re-vectoring kernel loops; the respective routines have been refactored to work on generic node "sections" (tuple of `Node`).
* `SCCAnnotate` converts `!$loki loop` and `!$loki routine` directives into `!$acc` equivalents and adds `private` clauses to vector loops and `!$acc data present` clauses to kernel routines.
* The `check_routine_pragmas` utility has been renamed `check_routine_sequential` and no longer inserts pragmas at all. It merely checks if a routine has been marked with `!$loki routine seq`. The second use case (check for repeated processing) is now handled explicitly in `SCCAnnotate` when annotating kernel routines (which cannot happen twice!).
* A general tidy-up of `SCCAnnotate` to make utility methods object-bound and re-use common ones in the driver processing code path. We still need the `block_dim`, but the `horizontal` is no longer needed in `SCCAnnotate`.
* `test_scc_vector.py` now also checks for the insertion of certain `!$loki loop` pragmas.

### CODE REVIEW

**Comment by mlange05** [comment]
Quick note: This is ready for review, but currently still fails ec-phys regression locally and will need some small fixes.

File: loki/transformations/single_column/vector.py [hunk 1]
```diff
@@ -264,48 +299,154 @@ class SCCRevectorTransformation(Transformation):
...
+        section : tuple of :any:`Node`
+            Code section in which to mark "seq loops".
+        """
+        for loop in FindNodes(ir.Loop).visit(section):
 
-        # Ensure we clone all body nodes, to avoid recursion issues
-        vector_loop = ir.Loop(variable=index, bounds=bounds, body=Transformer().visit(section))
+            # Skip loops explicitly marked with `!$loki/claw nodep`
+            if loop.pragma and any('nodep' in p.content.lower() for p in as_tuple(loop.pragma)):
+                continue
```
Comment (reuterbal) [inline loki/transformations/single_column/vector.py:366]: I think there's currently no test that covers this functionality.

File: loki/transformations/tests/test_utilities.py [hunk 2]
```diff
@@ -537,12 +537,10 @@ def test_transform_utilites_check_routine_pragmas(frontend, tmp_path):
     i = i + 1
   end subroutine test_acc_vec
 
-end module test_check_routine_pragmas_mod
+end module test_check_routine_sequential_mod
 """
     module = Module.from_source(fcode, frontend=frontend, xmods=[tmp_path])
 
-    # TODO: This utility needs some serious clean-up, so we're just testing
-    # the bare basics here and promise to do better next time ;)
```
Comment (reuterbal) [inline loki/transformations/tests/test_utilities.py:545]: Deleting this must have felt good 😏

**Review by reuterbal — APPROVED** [review]
Thanks, this is a really nice clean-up of that processing step that will make subsequent developments much easier and more flexible, and improve the ways how we can interact with and guide transformation pipelines manually when necessary.

One minor comment, nothing that strictly needs an action now.

File: loki/transformations/tests/test_utilities.py (hunk 2 above)
Comment (mlange05) [reply loki/transformations/tests/test_utilities.py:545]: No comment 😏

**Comment by mlange05** [comment]
Thanks for the review @reuterbal . However, I just confirmed that this still fails H24-dev locally, so will need to have another look. I'll remove the merge-ready label and rebase. Will ping once fixed.

**Comment by mlange05** [comment]
Small integration note: I've needed to rebase manually, as one of the changes in this (the explicit removal of vector-section objects/labels) was implement in a similar, but separate implementation in a recent PR (#328 ). I've also found the bugs that were causing local H24-dev to fail, and added fixes+tests for those corner cases (top two commits). I think I'm happy with this now (confirmed to work locally with H24 ecphys regression), so please have another look and merge once ready.

File: loki/transformations/single_column/tests/test_scc.py [hunk 3]
```diff
@@ -776,20 +774,57 @@ def test_scc_base_routine_seq_pragma(frontend, horizontal):
...
+    subroutine some_kernel(n)
+       implicit none
+       ! Scalars should not show up in `!$acc data` clause
+       integer, intent(inout) :: n
+!$loki routine seq
+       integer :: k
+
+       k = n
+       do k=1, 3
+          n = k + 1.
```
Comment (reuterbal) [inline loki/transformations/single_column/tests/test_scc.py:808]: Should this not be an integer literal `1`?

**Review by reuterbal — APPROVED** [review]
Thanks, looks all good to me!

File: loki/transformations/single_column/tests/test_scc.py (hunk 3 above)
Comment (mlange05) [reply loki/transformations/single_column/tests/test_scc.py]: Indeed, it should. Fixed now
